File: recamera-intellisense-mcp/src/api/video.rs
//! Video stream encode settings (`/video/{0,1}/encode`).

use anyhow::{bail, Result};
use serde_json::Value;

use crate::api::expect_ok;
use crate::api_client::ApiClient;
use crate::types::{DeviceRecord, SetVideoEncodeParams, VideoEncode, VideoEncodeSetResult};

fn stream_key_id(stream: &str) -> Result<(String, i32)> {
    let key = stream.trim().to_lowercase();
    let id = match key.as_str() {
        "main" => 0,
        "sub" => 1,
        _ => bail!("stream must be 'main' or 'sub'; got {stream:?}"),
    };
    Ok((key, id))
}

fn encode_path(stream: &str) -> Result<String> {
    let (_, id) = stream_key_id(stream)?;
    Ok(format!("/cgi-bin/entry.cgi/video/{id}/encode"))
}

/// Encode parameters of one stream (`main` = 0, `sub` = 1).
pub async fn get_video_encode(
    client: &ApiClient,
    device: &DeviceRecord,
    stream: &str,
) -> Result<VideoEncode> {
    let (key, _) = stream_key_id(stream)?;
    let d = client.get_json(device, &encode_path(stream)?, None).await?;
    let text = |k: &str| d.get(k).and_then(|v| v.as_str()).unwrap_or("").to_string();
    Ok(VideoEncode {
        stream: key,
        stream_type: text("sStreamType"),
        enabled: d.get("iEnabled").and_then(|v| v.as_i64()).unwrap_or(0) != 0,
        codec: text("sOutputDataType"),
        resolution: text("sResolution"),
        frame_rate: text("sFrameRate"),
        gop: d.get("iGOP").and_then(|v| v.as_i64()).unwrap_or(0) as i32,
        rc_mode: text("sRCMode"),
        rc_quality: text("sRCQuality"),
        max_rate: d.get("iMaxRate").and_then(|v| v.as_i64()).unwrap_or(0) as i32,
    })
}

const CODECS: [&str; 2] = ["H.264", "H.265"];
const RC_MODES: [&str; 2] = ["CBR", "VBR"];
const RC_QUALITIES: [&str; 4] = ["highest", "high", "medium", "low"];

fn norm_resolution(value: &str) -> Result<String> {
    let text = value.to_lowercase().replace('x', "*");
    let (w, h) = text
        .split_once('*')
        .and_then(|(a, b)| Some((a.parse::<i64>().ok()?, b.parse::<i64>().ok()?)))
        .ok_or_else(|| anyhow::anyhow!("resolution must be \"WxH\" within 384*384 ~ 3840*2160"))?;
    if !(384..=3840).contains(&w) || !(384..=2160).contains(&h) {
        bail!("resolution must be \"WxH\" within 384*384 ~ 3840*2160");
    }
    Ok(format!("{w}*{h}"))
}

/// Build the wire payload + the friendly verify map. Pure for unit tests.
fn build_encode_payload(
    params: &SetVideoEncodeParams,
) -> Result<(serde_json::Map<String, Value>, serde_json::Map<String, Value>)> {
    let mut payload = serde_json::Map::new();
    let mut checks = serde_json::Map::new();
    if let Some(codec) = &params.codec {
        let codec = codec.to_uppercase();
        if !CODECS.contains(&codec.as_str()) {
            bail!("codec must be one of {CODECS:?}");
        }
        payload.insert("sOutputDataType".into(), codec.clone().into());
        checks.insert("codec".into(), codec.into());
    }
    if let Some(resolution) = &params.resolution {
        let norm = norm_resolution(resolution)?;
        payload.insert("sResolution".into(), norm.clone().into());
        checks.insert("resolution".into(), norm.into());
    }
    if let Some(fps) = params.frame_rate {
        if !(1..=120).contains(&fps) {
            bail!("frame_rate must be 1~120");
        }
        payload.insert("sFrameRate".into(), fps.to_string().into()); // numeric STRING on the wire
        checks.insert("frame_rate".into(), fps.to_string().into());
    }
    if let Some(gop) = params.gop {
        if !(1..=120).contains(&gop) {
            bail!("gop must be 1~120");
        }
        payload.insert("iGOP".into(), gop.into());
        checks.insert("gop".into(), gop.into());
    }
    if let Some(rc_mode) = &params.rc_mode {
        let mode = rc_mode.to_uppercase();
        if !RC_MODES.contains(&mode.as_str()) {
            bail!("rc_mode must be one of {RC_MODES:?}");
        }
        payload.insert("sRCMode".into(), mode.clone().into());
        checks.insert("rc_mode".into(), mode.into());
    }
    if let Some(rc_quality) = &params.rc_quality {
        let quality = rc_quality.to_lowercase();
        if !RC_QUALITIES.contains(&quality.as_str()) {
            bail!("rc_quality must be one of {RC_QUALITIES:?}");
        }
        payload.insert("sRCQuality".into(), quality.clone().into());
        checks.insert("rc_quality".into(), quality.into());
    }
    if let Some(max_rate) = params.max_rate {
        if !(3..=65536).contains(&max_rate) {
            bail!("max_rate must be 3~65536");
        }
        payload.insert("iMaxRate".into(), max_rate.into());
        checks.insert("max_rate".into(), max_rate.into());
    }
    if let Some(enabled) = params.enabled {
        payload.insert("iEnabled".into(), (enabled as i64).into());
        checks.insert("enabled".into(), enabled.into());
    }
    if payload.is_empty() {
        bail!(
            "nothing to change: pass codec, resolution, frame_rate, gop, rc_mode, \
             rc_quality, max_rate, or enabled"
        );
    }
    Ok((payload, checks))
}

/// Update encode parameters; only passed fields change, and the result is
/// verified by re-reading the device. Applying briefly re-inits the encoder.
pub async fn set_video_encode(
    client: &ApiClient,
    device: &DeviceRecord,
    params: &SetVideoEncodeParams,
) -> Result<VideoEncodeSetResult> {
    let (stream, _) = stream_key_id(params.stream.as_deref().unwrap_or("main"))?;
    let (payload, checks) = build_encode_payload(params)?;
    let resp = client
        .post_json(
            device,
            &encode_path(&stream)?,
            None,
            Some(&Value::Object(payload)),
        )
        .await?;
    expect_ok(&resp, "set video encode")?;
    let after = get_video_encode(client, device, &stream).await?;
    let after_value = serde_json::to_value(&after)?;
    let mismatched: Vec<String> = checks
        .iter()
        .filter(|(k, v)| {
            after_value
                .get(k.as_str())
                .map(|got| got.to_string().trim_matches('"') != v.to_string().trim_matches('"'))
                .unwrap_or(true)
        })
        .map(|(k, v)| format!("{k}: wanted {v}, got {:?}", after_value.get(k.as_str())))
        .collect();
    if !mismatched.is_empty() {
        bail!("set video encode did not apply: {}", mismatched.join(", "));
    }
    Ok(VideoEncodeSetResult {
        changed: true,
        stream,
        applied: Value::Object(checks),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn params() -> SetVideoEncodeParams {
        SetVideoEncodeParams {
            device_name: String::new(),
            stream: None,
            codec: None,
            resolution: None,
            frame_rate: None,
            gop: None,
            rc_mode: None,
            rc_quality: None,
            max_rate: None,
            enabled: None,
        }
    }

    #[test]
    fn nothing_to_change_rejected() {
        assert!(build_encode_payload(&params()).is_err());
    }

    #[test]
    fn wire_types_and_normalization() {
        let mut p = params();
        p.codec = Some("H.265".into());
        p.resolution = Some("1920x1080".into());
        p.frame_rate = Some(25);
        p.rc_mode = Some("vbr".into());
        p.rc_quality = Some("HIGH".into());
        p.enabled = Some(true);
        let (payload, checks) = build_encode_payload(&p).unwrap();
        assert_eq!(payload["sFrameRate"], Value::from("25")); // string
        assert_eq!(payload["iEnabled"], Value::from(1)); // int
        assert_eq!(payload["sResolution"], Value::from("1920*1080"));
        assert_eq!(checks["rc_mode"], Value::from("VBR"));
        assert_eq!(checks["rc_quality"], Value::from("high"));
    }

    #[test]
    fn codec_case_normalized() {
        let mut p = params();
        p.codec = Some("h.265".into());
        let (payload, _) = build_encode_payload(&p).unwrap();
        assert_eq!(payload["sOutputDataType"], Value::from("H.265"));
    }

    #[test]
    fn out_of_range_rejected() {
        for (f, v) in [("frame_rate", 0), ("frame_rate", 121), ("gop", 0), ("max_rate", 2)] {
            let mut p = params();
            match f {
                "frame_rate" => p.frame_rate = Some(v),
                "gop" => p.gop = Some(v),
                _ => p.max_rate = Some(v),
            }
            assert!(build_encode_payload(&p).is_err(), "{f}={v}");
        }
        let mut p = params();
        p.codec = Some("VP9".into());
        assert!(build_encode_payload(&p).is_err());
        let mut p = params();
        p.resolution = Some("100*100".into());
        assert!(build_encode_payload(&p).is_err());
    }
}

File: skills/recamera-local/scripts/recamera_local/video.py
"""Video stream encode settings (``/video/{0,1}/encode``)."""

from __future__ import annotations

from typing import Any, Dict, Optional

from . import _local as http
from ._errors import RecameraError

__all__ = ["get_video_encode", "set_video_encode"]

PATH_ENCODE = "/cgi-bin/entry.cgi/video/{}/encode"

_STREAM_IDS = {"main": 0, "sub": 1}

_CODECS = ("H.264", "H.265")
_RC_MODES = ("CBR", "VBR")
_RC_QUALITIES = ("highest", "high", "medium", "low")


def get_video_encode(*, stream: str = "main") -> Dict[str, Any]:
    """Encode parameters for `stream` (``"main"``|``"sub"``):
    ``{stream, stream_type, enabled, codec, resolution, frame_rate, gop,
    rc_mode, rc_quality, max_rate(kbps)}``."""
    key = str(stream).strip().lower()
    if key not in _STREAM_IDS:
        raise ValueError(
            f"stream must be one of {sorted(_STREAM_IDS)}; got {stream!r}"
        )
    d = http.get_json(PATH_ENCODE.format(_STREAM_IDS[key]))
    if not isinstance(d, dict):
        raise RecameraError("get video encode failed: unexpected payload")
    return {
        "stream": key,
        "stream_type": d.get("sStreamType", ""),
        "enabled": bool(d.get("iEnabled", 0)),
        "codec": d.get("sOutputDataType", ""),
        "resolution": d.get("sResolution", ""),
        "frame_rate": d.get("sFrameRate", ""),
        "gop": int(d.get("iGOP", 0)),
        "rc_mode": d.get("sRCMode", ""),
        "rc_quality": d.get("sRCQuality", ""),
        "max_rate": int(d.get("iMaxRate", 0)),
    }


def _int_field(name: str, value: Any, lo: int, hi: int) -> int:
    try:
        ivalue = int(value)
    except (TypeError, ValueError):
        raise ValueError(f"{name} must be an integer {lo}~{hi}") from None
    if isinstance(value, bool) or (isinstance(value, float)
                                   and value != ivalue) \
            or not lo <= ivalue <= hi:
        raise ValueError(f"{name} must be an integer {lo}~{hi}")
    return ivalue


def _norm_resolution(value: Any) -> str:
    text = str(value or "").lower().replace("x", "*")
    try:
        w_s, h_s = text.split("*", 1)
        w, h = int(w_s), int(h_s)
    except (ValueError, AttributeError):
        raise ValueError(
            "resolution must be \"WxH\" within 384*384 ~ 3840*2160") from None
    if not (384 <= w <= 3840 and 384 <= h <= 2160):
        raise ValueError("resolution must be \"WxH\" within 384*384 ~ 3840*2160")
    return f"{w}*{h}"


def set_video_encode(
    *,
    stream: str = "main",
    codec: Optional[str] = None,
    resolution: Optional[str] = None,
    frame_rate: Optional[Any] = None,
    gop: Optional[Any] = None,
    rc_mode: Optional[str] = None,
    rc_quality: Optional[str] = None,
    max_rate: Optional[Any] = None,
    enabled: Optional[bool] = None,
) -> Dict[str, Any]:
    """Update encode parameters for `stream`; only the fields passed are
    changed (the rest keep their stored values), and the result is verified
    by re-reading the device. Changing these briefly re-inits the encoder.
    """
    key = str(stream).strip().lower()
    if key not in _STREAM_IDS:
        raise ValueError(
            f"stream must be one of {sorted(_STREAM_IDS)}; got {stream!r}")
    payload: Dict[str, Any] = {}
    checks: Dict[str, Any] = {}
    if codec is not None:
        codec = str(codec).upper()
        if codec not in _CODECS:
            raise ValueError(f"codec must be one of {list(_CODECS)}")
        payload["sOutputDataType"] = codec
        checks["codec"] = codec
    if resolution is not None:
        payload["sResolution"] = _norm_resolution(resolution)
        checks["resolution"] = payload["sResolution"]
    if frame_rate is not None:
        fps = _int_field("frame_rate", frame_rate, 1, 120)
        payload["sFrameRate"] = str(fps)  # wire type is a numeric string
        checks["frame_rate"] = str(fps)
    if gop is not None:
        payload["iGOP"] = _int_field("gop", gop, 1, 120)
        checks["gop"] = payload["iGOP"]
    if rc_mode is not None:
        rc_mode = str(rc_mode).upper()
        if rc_mode not in _RC_MODES:
            raise ValueError(f"rc_mode must be one of {list(_RC_MODES)}")
        payload["sRCMode"] = rc_mode
        checks["rc_mode"] = rc_mode
    if rc_quality is not None:
        rc_quality = str(rc_quality).lower()
        if rc_quality not in _RC_QUALITIES:
            raise ValueError(
                f"rc_quality must be one of {list(_RC_QUALITIES)}")
        payload["sRCQuality"] = rc_quality
        checks["rc_quality"] = rc_quality
    if max_rate is not None:
        payload["iMaxRate"] = _int_field("max_rate", max_rate, 3, 65536)
        checks["max_rate"] = payload["iMaxRate"]
    if enabled is not None:
        payload["iEnabled"] = 1 if enabled else 0
        checks["enabled"] = bool(enabled)
    if not payload:
        raise ValueError(
            "nothing to change: pass codec, resolution, frame_rate, gop, "
            "rc_mode, rc_quality, max_rate, or enabled")
    http.expect_ok(
        http.post_json(PATH_ENCODE.format(_STREAM_IDS[key]), payload=payload),
        "set video encode")
    after = get_video_encode(stream=key)
    mismatched = {
        k: {"wanted": v, "got": after.get(k)}
        for k, v in checks.items()
        if str(after.get(k)) != str(v)
    }
    if mismatched:
        raise RecameraError(
            f"set video encode did not apply: {mismatched}")
    return {"changed": True, "stream": key, "applied": checks}


COMMANDS = {
    "get_video_encode": get_video_encode,
    "set_video_encode": set_video_encode,
}
